### ISSUE TITLE
New combine operator

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#298]
Introduce a new `combine` operator that provides the following benefits over the existing `spread` operator. 

* More intuitive name.
* Stream item combinations as soon as they are received. Currently `spread` needs to convert the channel operand to a list object. This implies that combinations can be emitted only after *all* items are previously collected.   
* Preserve nested list objects. Fix #241. 
* Allow combinations only between items having a shared key. Fix #253. 

For example: 
```
numbers = Channel.from(1,2,3)
words = Channel.from('hello', 'ciao')
numbers
    .combine(words)
    .println()

# outputs
[1, hello]
[2, hello]
[3, hello]
[1, ciao]
[2, ciao]
[3, ciao]
```

The following example shows how combine items by a matching key: 
```
left = Channel.from(['A',1], ['B',2], ['A',3])
right = Channel.from(['B','x'], ['B','y'], ['A','z'], ['A', 'w'])

left
    .combine(right, by: 0)
    .println()

# outputs
[A, 1, z]
[A, 3, z]
[A, 1, w]
[A, 3, w]
[B, 2, x]
[B, 2, y]
```

**Comment by mfoll** [comment]
> Introduce a new `collect` operator that provides the following benefits over the existing spread operator.

Typo here: should be `combine` instead of `collect`

**Comment by pditommaso** [comment]
Oops, fixed!

**Comment by msmootsgi** [comment]
So glad it's `combine` and not `collect`!

**Comment by pditommaso** [comment]
But there's also `collect` #297. Hope that's fine :)

**Comment by msmootsgi** [comment]
Oh, I see.  I'm just afraid that channel mapping functions that use groovy's `collect` on channel data like this:
```
.flatMap{ it.withIndex().collect{ pe, ind -> tuple(it.name, ind, pe) } }
```
will get even more confusing if there's now also a channel `collect`.

**Comment by pditommaso** [comment]
IMO that's fine because it's coherent with the semantic of the operation and the fact of using a OOP on which you have the same method applied to different object types.

**Comment by msmootsgi** [comment]
Yeah, now that I look at `collect` I see that I was misinterpreting the semantics of `collect` and that it (almost) always returns a `List`.  I was thinking it was a functor.  So, good change and sorry for the distraction!

**Comment by pditommaso** [comment]
Included in version `0.24.0`
